File: lib/magic.go
package lib

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
)

type Card struct {
	Name      string `json:"name"`
	Layout    string `json:"layout"`
	ImageUris struct {
		Small      string `json:"small"`
		Normal     string `json:"normal"`
		Large      string `json:"large"`
		Png        string `json:"png"`
		ArtCrop    string `json:"art_crop"`
		BorderCrop string `json:"border_crop"`
	} `json:"image_uris"`
	PrintSearchUri string `json:"prints_search_uri"`
	CardFaces      []struct {
		Object         string   `json:"object"`
		Name           string   `json:"name"`
		ManaCost       string   `json:"mana_cost"`
		TypeLine       string   `json:"type_line"`
		OracleText     string   `json:"oracle_text"`
		Colors         []string `json:"colors"`
		Loyalty        string   `json:"loyalty"`
		Artist         string   `json:"artist"`
		ArtistID       string   `json:"artist_id"`
		IllustrationID string   `json:"illustration_id"`
		ImageUris      struct {
			Small      string `json:"small"`
			Normal     string `json:"normal"`
			Large      string `json:"large"`
			Png        string `json:"png"`
			ArtCrop    string `json:"art_crop"`
			BorderCrop string `json:"border_crop"`
		} `json:"image_uris"`
	} `json:"card_faces"`
}

type Prints struct {
	Object     string `json:"object"`
	TotalCards int    `json:"total_cards"`
	HasMore    bool   `json:"has_more"`
	NextPage   string `json:"next_page"`
	Data       []struct {
		ID        string `json:"id"`
		Name      string `json:"name"`
		Layout    string `json:"layout"`
		ImageUris struct {
			Small      string `json:"small"`
			Normal     string `json:"normal"`
			Large      string `json:"large"`
			Png        string `json:"png"`
			ArtCrop    string `json:"art_crop"`
			BorderCrop string `json:"border_crop"`
		} `json:"image_uris"`
		CardFaces []struct {
			Name      string `json:"name"`
			ImageUris struct {
				Small      string `json:"small"`
				Normal     string `json:"normal"`
				Large      string `json:"large"`
				Png        string `json:"png"`
				ArtCrop    string `json:"art_crop"`
				BorderCrop string `json:"border_crop"`
			} `json:"image_uris"`
		} `json:"card_faces"`
		SetName string `json:"set_name"`
	} `json:"data"`
}

func Split(r rune) bool {
	return r == '(' || r == ')'
}

func SetDive(setname string, apiuri string) string {
	imagetoreturn := ""

	response, err := http.Get(apiuri)
	if err != nil {
		fmt.Print(err.Error())
		return "error"
	}

	responseData, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Print(err.Error())
		return "error parsing set details"
	}

	var prints Prints
	json.Unmarshal(responseData, &prints)

	for _, element := range prints.Data {
		if strings.EqualFold(setname, element.SetName) { //make fuzzy
			if element.Layout == "transform" {
				cardString := ""
				for _, face := range element.CardFaces {
					cardString += face.ImageUris.Large + " "
				}
				imagetoreturn = cardString
			} else {
				imagetoreturn = element.ImageUris.Large
			}
		}
	}

	if prints.HasMore {
		imagetoreturn = SetDive(setname, prints.NextPage)
	}

	if imagetoreturn == "" {
		return fmt.Sprintf("Card does not exist for set %s", setname)
	} else {
		return imagetoreturn
	}
}

func GetCard(cardname []string) string {
	commandValue := strings.Join(cardname[1:], " ")
	setSplit := strings.FieldsFunc(commandValue, Split)
	name := setSplit[0]

	response, err := http.Get(fmt.Sprintf("https://api.scryfall.com/cards/named?fuzzy=%s", name))

	if err != nil {
		fmt.Print(err.Error())
		return "error"
	}

	responseData, err := ioutil.ReadAll(response.Body)
	if err != nil {
		fmt.Print(err.Error())
		return "error parsing response"
	}

	var card Card
	json.Unmarshal(responseData, &card)

	//specific set requested. Time to search
	if len(setSplit) > 1 {
		fmt.Println(setSplit[1])
		imagetoreturn := SetDive(setSplit[1], card.PrintSearchUri)
		return imagetoreturn
	}

	if card.Layout == "transform" {
		cardString := ""
		for _, element := range card.CardFaces {
			cardString += element.ImageUris.Large + " "
		}
		return cardString
	} else {
		return card.ImageUris.Large
	}
}

File: src/magic.py
import requests


def get_card(user_input):
    input_split = user_input.split(" ")[1:]
    input_split = ' '.join(input_split)
    set_name = ""

    if "(" in input_split and ")" in input_split:
        if len(input_split) > input_split.find(")")+1:
            return "Too many arguments. Please use the format: !card <card_name> (<set_name>)"
        card_name = input_split[0:input_split.find("(")]
        set_name = input_split[input_split.find("(")+1:input_split.find(")")]
        print(set_name)
    else:
        card_name = input_split

    r = requests.get('https://api.scryfall.com/cards/named?fuzzy=' + card_name)
    return_response = r.json()

    if return_response["object"] == "error":
        return return_response["details"]
    
    if set_name:
        return set_dive(set_name, return_response["prints_search_uri"])
    
    if return_response["layout"] == "transform" or return_response["layout"] == "modal_dfc":
        card_string = ""
        for face in return_response["card_faces"]:
            card_string += face["image_uris"]["large"] + " "
        return card_string
    else:
        return return_response["image_uris"]["large"]
        

def set_dive(set_name, api_uri):
    images_to_return = ""
    
    while True:
        r = requests.get(api_uri)
        request_response = r.json()
        matches = [x for x in request_response["data"] if x["set"] == set_name.lower() or set_name == x["set_name"]]

        if matches:
            for card in matches:
                if card["layout"] == "transform" or card["layout"] == "modal_dfc":
                    card_string = ""
                    for face in card["card_faces"]:
                        card_string += face["image_uris"]["large"] + " "
                    images_to_return += card_string
                else:
                    images_to_return += card["image_uris"]["large"] + " "
        
        if not request_response["has_more"]:
            break
        else:
            api_uri = request_response["next_page"]

    if not images_to_return:
        return "Card does not exist for set " + set_name
    else:
        return images_to_return


def set_lookup(set_name):
    r = requests.get('https://api.scryfall.com/sets')
    request_response = r.json()
    matches = [x for x in request_response["data"] if x["parent_set_code"] == set_name.lower()]
    if matches:
        return "yes"